packages/react-spinbutton/Spec.md

```diff
@@ -2,62 +2,381 @@
 
 - _Migration from v8_
+
+1. Ensure `value` prop is a number, not a string
+2. Replace `onIncrement` and `onDecrement` callbacks with `onChange`.
+   1. Increment/decrement logic can be handled by comparing `data.value` and `data.prevValue` in the `onChange` callback.
+3. Update `onChange` callback to handle new signature.
+4. Remove `onValidate` callback.
+5. Change ARIA props.
+   1. Change non-standard props to standard ones (e.g., `ariaLabel` to `aria-label`).
+   2. Remove `ariaPositionInSet`, `ariaSetSize`, `ariaValueNow`, `ariaValueText`. The first two are not relevant for a spinbutton and the latter two are internal implementation details managed by the component.
+
 - _Migration from v0_
 
+Not applicable as v0 does not implement this component or one like it.
+
 ## Behaviors
 
-_Explain how the component will behave in use, including:_
+`SpinButton`'s `value` prop is always a number, in contrast to the v8 implementation that gave `value` a string typez. `SpinButton`s manipulate numeric values and making `value` a number aligns it with the other related props: `min`, `max` and `step`. `SpinButton`'s `value` is always displayed as a string which is determined with the `formatter()` function. By default this function stringifies `value` but it can do more sophisticated formatting like apply currency formatting. The inverse of `formatter()` is `parser()` which takes the stringified display value and converts it back to a number. The default implementation calls `parseFloat()` on the display value. `parser()` is called when a user directly types a value into `SpinButton`'s `<input>` element so this value can be converted to a number for computation. If `parser()` returns a non-number value (e.g., `NaN`, `undefined`, `null`) `value` will not be changed.
+
+`formatter()` and `parser()` allow users to hook into `SpinButton`'s display behavior without having to reimplement features like clamping `value` between `min` and `max` which is required when adding [custom suffixes in v8](https://codepen.io/seanms/pen/QWqpQWp).
+
+`SpinButton` does not allow values less than `min` or greater than `max` when these props are specified. When using the step buttons/arrow keys the value will not step outside the min/max bounds. If a value is typed into `<input>` that is outside the min/max range it will be clamped to the relevant bound, that is values less than `min` will be set to `min` and values greater than `max` will be set to `max`.
+
+Aside from min/max range clamping `SpinButton` does not currently implement any input validation.
+
+No error states are currently implemented.
+
+`SpinButton`'s `onChange` callback is invoked every time a change is committed. A change is committed when:
+
+1. A step button is activated
+2. When a user presses the `Up` or `Down` arrow keys while focused on the component.
+3. When a user presses the `Enter` key while focused on the `<input>`
+4. When the `blur` event is fired from the `<input>`
```

Comment:
Can you please clarify why this is required? We should try to mimic behaviour of native controls and native `input` does not emit `change` event for `blur`. IMO, we should call `onChange` only when `value` actually changes.

Reply:
blur would only cause a change event if the value in the field has changed.

Reply:
There needs to be some way to commit a value that has been changed by directly editing the text field. When a user is focused on the text `<input>` and directly entering a value we want to relax the normal min-max clamping rules and not fire any change events because they will be in an intermediate state as they type.

For example, say I have a `SpinButton` with a `min` of 5 and a `max` of 15 with a current value of 10. I focus on the `<input>` and place my cursor at the end of input, press backspace and now the input displays "1". Next I type "2" so the input now displays "12".

In this case we cannot apply clamping rules or we'll prevent a user from typing a valid value. If we fire a change event while in this intermediate state we will be reporting a value outside the min-max range. I feel that when `min` and `max` are set on `SpinButton` the `onChange` event should never report a value outside of the min-max range.

Using the blur event is a nice way for the user to signal "I'm done editing this field". That said, I'm happy to discuss alternatives.